### PR TITLE
Sourcecred Gitcoin Plugin 

### PR DESCRIPTION
**Pull request by psudoanon**
[Architecture Specification can be found here](https://gist.github.com/psudoanon/3847235ad13835145f14d76c76b47d6f)
[Instructions for running with Gitcoin can be found here](https://gist.github.com/psudoanon/24ff5f3e93f887fa8c5ca3fb0f68209c)

Let me know what you guys think :)

### CODE REVIEW

**Comment by decentralion** [comment]
Hey @psudoanon, took a first pass and it looks reasonable. can you please get CI passing (first step is fixing the flow issues). `yarn test` should pass, it's ok if test_full fails b.c. it's coming from a fork.

Before this merges, I'd like to sync with @owocki to understand who on the Gitcoin side will be responsible for maintaining this, would also like to see a generated cred instance with Gitcoin's prod data so we can dig into the cred scores a big (i think this is all on @owocki based on the other thread).

**Comment by decentralion** [comment]
BTW, you shouldn't need to include the ledger from the test instance. an empty ledger files (or one that just contains []) should be fine.

**Comment by decentralion** [comment]
Also: @hammadj and I were planning on removing the docker container. Can the Gitcoin usecase just go thru node/npm like everyone else?

**Comment by psudoanon** [comment]
@decentralion Hey, sorry about the flow issues, those should all be fixed now :)

I started a conversation with @owocki regarding what direction we want to go in regarding the Docker container so hopefully we'll have some direction on that soon. I'll work on updating the documentation after that.

**Comment by psudoanon** [comment]
From a chat with @owocki 

Screenshot_20200903-153254_Brave

**Comment by decentralion** [comment]
Yeah, I really don't think losing the docker container should matter for this purpose--the docker container was useful when we weren't on npm and the only ways to use SourceCred were to use the docker image or to build from source. But now it's just a matter of doing `npm install sourcecred` and then `node sourcecred load`, etc. cc @owocki

**Comment by owocki** [comment]
hiya; just got a sec to run this, loading up the docker instructions.    will post any feedback/issues i run into on the gist

i dont have strong feelings on the docker archtiecture direction; defer to @decentralion bc its their repo :)

**Comment by owocki** [comment]
hey @psudoanon  thanks for your patience as i was getting to this. i left a comment on the instructions about a blocking issue i encountered.

in leiu of me being able to actualy run it, would you be able to tell me what the output from this command looks like?  i basically just want to get a list of usernames and how much cred/grain they got.  im assuming thats what the output of `docker run -p 8080:8080 sourcecred/sourcecred gitcoin` is?

**Comment by decentralion** [comment]
@owocki the way to view Cred scores within a repo is to set up an instance (e.g. https://github.com/sourcecred/template-instance as a template, https://github.com/sourcecred/cred as a real example). then inside the instance, you can run `yarn serve` to spin up a web server that shows the cred scores, the same UI as http://cred.sourcecred.io 

Not sure how this would all work with the docker container involved--my advice would be to just build sourcecred from the fork (`yarn build`) and then use it, following the instructions from the template instance. You'll need to use the locally built sourcecred binary rather than the one pulled from npm though. (@psudoanon if you want to make this a bit easier for @owocki you can publish gitcoin-sourcecred-demo on npm)

**Comment by decentralion** [comment]
Also: we have dev meetings every Monday at 12pm PT (see: https://sourcecred.io/calendar). they're hosted on our Discord. you're both welcome to come by if you want to chat with us synchronously about the integration. @psudoanon, it would be great to see you and get to say hi given the work you've been doing on the project!

**Comment by owocki** [comment]
thanks @decentralion - will give that a shot once i get past the build issues + added the dev meeting to my calendar.  @psudoanon i could make it if u can

**Comment by psudoanon** [comment]
@owocki Yup, Monday works for me 👍 

I forked the template and published an npm module with the gitcoin plugin enabled, you can clone it from here: https://github.com/psudoanon/gitcoin-sourcecred-demo

Getting it working should be just updating the `config/plugins/sourcecred/gitcoin/config.json` and running `npm install`, after that you should be able to just run the commands in the `package.json`

Let me know if you run into any issues

**Comment by owocki** [comment]
just sent you some ETH to pay on time.  i still cant seem to get it to run but im also deep in grants round 7 stuff rn.  

pumped to get some actual rpodcution data results soon :)

**Comment by psudoanon** [comment]
@owocki awesome thanks!

Let me know what your issues are - do you think you'll have time this week for a sync up? Let me know what your environment is (node version, operating system, etc.) and the errors you're having and I'll see if I can reproduce on my end.

**Comment by owocki** [comment]
i left a comment on the gist letting you know where my error was. https://gist.github.com/psudoanon/24ff5f3e93f887fa8c5ca3fb0f68209c

**Comment by psudoanon** [comment]
@owocki Take a look at the readme in this [repository](https://github.com/psudoanon/gitcoin-sourcecred-demo) and clone it to get a local version of SourceCred with the Gitcoin plugin enabled running. The gist is out of date.

**Comment by owocki** [comment]
fishing for more support from my team to push this forward right now..

**Comment by owocki** [comment]
spec is here https://github.com/gitcoinco/web/issues/6813 and here https://github.com/gitcoinco/web/issues/6690

**Comment by chibie** [comment]
@psudoanon @decentralion 

i followed these steps in trying to test sourcecred-gitcoin without docker as suggested. i had gitcoin postgres docker container already running.

1. `git clone https://github.com/psudoanon/gitcoin-sourcecred-demo.git` ✅ 
2. `cd gitcoin-sourcecred-demo` and `yarn` ✅ 
3. `yarn load` ✅ 
4. `yarn graph` ✅ 
5. `yarn score` ❌ 

step 5 failed and returned the error below

```
yarn run v1.22.4
$ sourcecred-gitcoin-demo score
  GO   score
Error: line 1 is not valid JSON: SyntaxError: Unexpected end of JSON input
    at eval (webpack:///./src/util/jsonLog.js?:23:551)
    at Array.forEach (<anonymous>)
    at Function.fromString (webpack:///./src/util/jsonLog.js?:23:463)
    at Function.parse (webpack:///./src/ledger/ledger.js?:173:97)
    at scoreCommand (webpack:///./src/cli/score.js?:23:595)
error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
```

yarn version: v1.22.4
operating system: macOS Catalina 10.15.6

is there something i'm doing wrong?

**Comment by topocount** [comment]
@chibie this file should be pretty in your instance and populated as it is in the template:

https://github.com/sourcecred/template-instance/blob/master/data/ledger.json

**Comment by topocount** [comment]
If that ledger.json file is present, please share a link to the instance you're testing against.

**Comment by decentralion** [comment]
@topocount: The repo being cloned (https://github.com/psudoanon/gitcoin-sourcecred-demo) does have a ledger already present. But AFAIK the `[]` has always been supported as a valid ledger.json. So I'm confused why we're seeing that failure. I wonder what version of sourcecred was being used by @psudoanon in the gitcoin demo

**Comment by topocount** [comment]
@chibie @decentralion @psudoanon I tried testing locally with this branch, but I've run into a series of issues gyp-building libpq on my ubuntu instance. I'll take another shot later.

I have a feeling ledger.json isn't empty by the time  `yarn score` is running. It'd be helpful to see what's in there.

**Comment by chibie** [comment]
@topocount @decentralion @owocki  i gave this another shot.

for some good weird reason, i no longer get the error reported earlier.

here's the changes in `ledger.json`

**before `yarn graph`**: `[]`

**after `yarn graph`**:

```
{"action":{"identity":{"address":"N\u0000sourcecred\u0000core\u0000IDENTITY\u0000GmA9SPu5cYYhBZ5D7WZSjw\u0000","aliases":[],"id":"GmA9SPu5cYYhBZ5D7WZSjw","name":"root","subtype":"USER"},"type":"CREATE_IDENTITY"},"ledgerTimestamp":1603790389865,"uuid":"aOdSG7BZWPwYrVsL3YpLCw","version":"1"}
{"action":{"alias":{"address":"N\u0000sourcecred\u0000gitcoin\u0000user\u0000localhost:8080\u0000root\u0000","description":"gitcoin/root"},"identityId":"GmA9SPu5cYYhBZ5D7WZSjw","type":"ADD_ALIAS"},"ledgerTimestamp":1603790389865,"uuid":"9hGQtKVepjAxwK7ZnQL8FA","version":"1"}
```

**after `yarn score`**: 

```
{"action":{"identity":{"address":"N\u0000sourcecred\u0000core\u0000IDENTITY\u0000GmA9SPu5cYYhBZ5D7WZSjw\u0000","aliases":[],"id":"GmA9SPu5cYYhBZ5D7WZSjw","name":"root","subtype":"USER"},"type":"CREATE_IDENTITY"},"ledgerTimestamp":1603790389865,"uuid":"aOdSG7BZWPwYrVsL3YpLCw","version":"1"}
{"action":{"alias":{"address":"N\u0000sourcecred\u0000gitcoin\u0000user\u0000localhost:8080\u0000root\u0000","description":"gitcoin/root"},"identityId":"GmA9SPu5cYYhBZ5D7WZSjw","type":"ADD_ALIAS"},"ledgerTimestamp":1603790389865,"uuid":"9hGQtKVepjAxwK7ZnQL8FA","version":"1"}
{"action":{"identity":{"address":"N\u0000sourcecred\u0000core\u0000IDENTITY\u0000J6556GIdBdoaSZdj0nLbDA\u0000","aliases":[],"id":"J6556GIdBdoaSZdj0nLbDA","name":"SourceCred","subtype":"PROJECT"},"type":"CREATE_IDENTITY"},"ledgerTimestamp":1603790480753,"uuid":"5w0qH3JWLCoPxyopPxnsOQ","version":"1"}
{"action":{"identityId":"J6556GIdBdoaSZdj0nLbDA","type":"TOGGLE_ACTIVATION"},"ledgerTimestamp":1603790480753,"uuid":"f4rnMlKC8VDhP8oF0fyefg","version":"1"}
```

so currently:

1. git clone https://github.com/psudoanon/gitcoin-sourcecred-demo.git ✅
2. cd gitcoin-sourcecred-demo and yarn ✅
3. yarn load ✅
4. yarn graph ✅
5. yarn score ✅

output files generated:
1. `graphs/sourcecred/gitcoin/graph.json`
2. `accounts.json`
3. `credResult.json`

**Comment by topocount** [comment]
Glad to hear it @chibie !

**Comment by topocount** [comment]
@decentralion @owocki what are the next steps here? I'm inclined to get this rebased onto master and ensure it still works, @chibie , if you're up for that. 

I remember a while ago Dandelion mentioning that this was planned to be merge into this repo's master branch, so it'd be great if we could outline what, if anything remains before that can be done. All I see in this thread is 

- [ ] @owocki needs to verify that Gitcoin's prod data looks correct when using the plugin

File: src/plugins/gitcoin/plugin.js
Comment (chibie) [inline src/plugins/gitcoin/plugin.js:71]: @psudoanon also require GITCOIN_USER_WHITELIST as an environmental variable.

**Review by chibie — CHANGES_REQUESTED** [review]

**Comment by decentralion** [comment]
Closing as replaced by #2452. Thanks @psudoanon for the work here! Once the plugin is getting used, let's make sure to flow Cred back to this PR.

**Pull request closed without merging**
